File: 12-lab/report1/lab2-vss.py
import random
from sympy import isprime
from Crypto.Util import number

# parameters setting
# prime q of bit-length 16, polynomial and shares in Fq
q = number.getPrime(16)
k = 1
while True:
    p = k * q + 1  # prime p ,q|p-1, commitments in Fp
    if isprime(p):
        break
    k = k + 1

# Compute elements of G_q = {h^k mod p | h in Z_p^*}
# |G_q| = q ,G_q is the unique q-order subgroup of Z_p.
G = []
for i in range(1, p):
    G.append(i**k % p)

G = list(set(G))
G.sort()
# print("G_q = ", G)
print("Order of G is " + str(len(G)) + " = q")

# Since the order of G_q is prime, any element of G_q except 1 is a generator
g = random.choice(list(filter(lambda g: g != 1, G)))  # g \in G_q
print("g = " + str(g))

# Not precise, but feasible.
"""
# y=u mod q
# 1、g^f(x) %p = g^{y %q} %p = g^u %p
# 2、(com[0]^x_0) * ... * (com[n-1]^x_n-1) %p = g^y %p = g^{u+tq} %p = g^u %p
if    g^q=1 mod p

g = random.randrange(2,p)
while(pow(g,q,p) != 1 ):
        g = random.randrange(2, p)
"""

print("q=", q, " g=", g, " p=", p)


def mod_inverse(a, m):
    m0, x0, x1 = m, 0, 1
    while a > 1:
        q = a // m
        m, a = a % m, m
        x0, x1 = x1 - q * x0, x0
    return x1 + m0 if x1 < 0 else x1


def coeff(degree, secret):
    coefficients = []
    coefficients += [random.randrange(0, q) for _ in range(degree - 1)]
    coefficients.append(secret)
    print(coefficients)
    return coefficients


def polynom(x, coefficients):
    point = 0
    for coefficient_index, coefficient_value in enumerate(coefficients[::-1]):
        point += x**coefficient_index * coefficient_value
    return point


def generate_shares(num_shares, degree, secret, g):
    coefficients = coeff(degree, secret)
    print("coefficients ", coefficients)
    shares = []
    xs = random.sample(range(1, q), num_shares)  # 采取  num_shares  个随机数
    commitments = []
    for i in range(degree):
        commitments.append(pow(g, coefficients[degree - 1 - i], p))
    for i in range(0, num_shares):
        x = xs[i]
        shares.append((x, polynom(x, coefficients) % q))
    return shares, commitments


def reconstruct_secret(shares):
    sums = 0
    for j, share_j in enumerate(shares):
        xj, yj = share_j
        prod = 1
        for i, share_i in enumerate(shares):
            xi, _ = share_i
            if i != j and (xi - xj) != 0:
                prod = (prod * ((-xi) * mod_inverse((xj - xi) % q, q))) % q
        prod = (prod * yj) % q
        sums = (sums + prod) % q
    return sums


def Verify(degree, num_shares, shares):
    for i in range(num_shares):
        Pi = 1
        for j in range(degree):
            Pi *= pow(commitments[j], pow(shares[i][0], j), p) % p

        if Pi % p == pow(g, shares[i][1], p):
            print(
                f"Verification successful for Participant {i}. The share is valid")
        else:
            print("Verification failed. The share is invalid.")


if __name__ == "__main__":
    # (3,5) sharing scheme
    threshold, num_shares = 3, 5
    degree = threshold
    secret = 1234  # Fq
    print(f"Original Secret: {secret}")

    # Phase I: Generation of shares
    shares, commitments = generate_shares(num_shares, degree, secret, g)

    print(f'Shares: {", ".join(str(share) for share in shares)}')
    print("commitments", commitments)

    # Phase II: Validation of Shares
    Verify(degree, num_shares, shares)

    # Phase II: Secret Reconstruction
    # Picking degree shares randomly for
    # reconstruction
    pool = random.sample(shares, degree)
    print(f'Combining shares: {", ".join(str(share) for share in pool)}')
    print(f"Reconstructed secret: {reconstruct_secret(pool)}")

File: 12-lab/report1/lab2-vss.cc
#include <NTL/ZZ.h>
#include <algorithm>
#include <cassert>
#include <iostream>
#include <iterator>
#include <random>
#include <set>
#include <vector>

using namespace std;
using namespace NTL;

using Share = pair<ZZ, ZZ>;

ZZ p, q, g;
size_t bits = 16;

long witness(const ZZ &n, const ZZ &x) {
  ZZ m, y, z;
  long j, k;

  if (x == 0)
    return 0;

  // compute m, k such that n-1 = 2^k * m, m odd:

  k = 1;
  m = n / 2;
  while (m % 2 == 0) {
    k++;
    m /= 2;
  }

  z = PowerMod(x, m, n); // z = x^m % n
  if (z == 1)
    return 0;

  j = 0;
  do {
    y = z;
    z = (y * y) % n;
    j++;
  } while (j < k && z != 1);

  return z != 1 || y != n - 1;
}

long prime_test(const ZZ &n, long t) {
  if (n <= 1)
    return 0;

  // first, perform trial division by primes up to 2000

  PrimeSeq s; // a class for quickly generating primes in sequence
  long p;

  p = s.next(); // first prime is always 2
  while (p && p < 2000) {
    if ((n % p) == 0)
      return (n == p);
    p = s.next();
  }

  // second, perform t Miller-Rabin tests

  ZZ x;
  long i;

  for (i = 0; i < t; i++) {
    x = RandomBnd(n); // random number between 0 and n-1

    if (witness(n, x))
      return 0;
  }

  return 1;
}

// b = a^-1 mod m
void mod_inverse(ZZ &b, ZZ a, ZZ m) {
  ZZ m0, x0, x1;
  m0 = m, x0 = 0, x1 = 1;
  while (a > 1) {
    ZZ q = a / m;
    ZZ a0 = a;
    a = m;
    m = a0 % m;

    ZZ x00 = x0;
    x0 = x1 - q * x0;
    x1 = x00;
  }
  b = x1 < 0 ? x1 + m0 : x1;
}

// f = v + a1x + a2x^2 ... % q
void gen_coeff(vector<ZZ> &coeffs, ZZ v, size_t t) {
  assert(coeffs.size() == 0);

  ZZ rnd;
  for (size_t i = 0; i < t - 1; ++i) {
    RandomBits(rnd, bits);
    rnd %= q;
    coeffs.emplace_back(rnd);
  }
  coeffs.emplace_back(v);
}

// f = f(x) % q
void evaluate_poly(ZZ &f, const vector<ZZ> &coeffs, const ZZ &x) {
  f = 0;
  ZZ m(1);

  // for (auto c : std::ranges::views::reverse(coeffs)) {
  for (auto it = coeffs.rbegin(); it != coeffs.rend(); ++it) {
    auto c = *it;
    f += m * c;
    m = m * x;
  }
  f %= q;
}

// genrate shares and commitments
void vss_share(vector<Share> &shares, vector<ZZ> &comms, const ZZ &secret,
               size_t share_num, size_t degree) {
  assert(shares.size() == 0);

  vector<ZZ> coeffs;
  // generator a `t-1` degree polynomial
  gen_coeff(coeffs, secret, degree);

  // cout << "coefficients: " << endl;
  // for (auto c : coeffs) {
  //   cout << "\t" << c << endl;
  // }
  // cout << endl;

  // make commitments
  for (size_t i = 0; i < degree; ++i) {
    comms.emplace_back(PowerMod(g, coeffs[degree - 1 - i], p));
  }

  // get n shares
  for (size_t i = 0; i < share_num; ++i) {
    ZZ x, f;

    // generate a random value
    RandomBits(x, bits);
    x %= q;

    // evaluate out f on x
    evaluate_poly(f, coeffs, x);
    shares.push_back(make_pair(x, f));
  }
}

// just lagrange...
void vss_recon(ZZ &v, vector<Share> shares) {

  size_t t = shares.size();
  v = 0;

  for (size_t j = 0; j < t; ++j) {
    auto &[xj, yj] = shares[j];
    ZZ prod(1);

    for (size_t i = 0; i < t; ++i) {
      auto xi = shares[i].first;
      if (i != j && (xi - xj) != 0) {
        ZZ tmp;
        mod_inverse(tmp, (xj - xi) % q, q);
        prod = (prod * ((-xi) * tmp)) % q;
      }
    }

    prod = (prod * yj) % q;
    v = (v + prod) % q;
  }
}

void vss_verify(const vector<Share> &shares, const vector<ZZ> comms,
            size_t share_num, size_t degree) {

  for (size_t i = 0; i < share_num; ++i) {
    ZZ Pi(1);
    for (size_t j = 0; j < degree; ++j) {
      Pi *= PowerMod(comms[j], power(shares[i].first, j), p);
    }

    if (Pi % p == PowerMod(g, shares[i].second, p)) {
      cout << "Verification successful for Participant " << i
           << ". The share is valid" << endl;
    } else {
      cout << "Verification failed for Participan " << i
           << ". The share is invalid." << endl;
    }
  }
}

void print_shares(vector<Share> shares) {
  for (auto &[x, f] : shares) {
    cout << "\t(" << x << ", " << f << ")" << endl;
  }
}

int main() {
  ZZ v;
  size_t n, t;

  // init paremeters: (n, t) share
  GenPrime(q, bits);
  // RandomBits(v, bits);
  // https://stackoverflow.com/questions/31386544/ntl-library-how-to-assign-a-big-integer-to-zz-p
  // v = conv<ZZ>("1234567890987654321234567890111222333444555666777888999");
  v = 1234;
  v %= q;
  // n = 15;
  // t = 10;
  n = 5;
  t = 3;

  // find prime p, q | p - 1, commitments in Fp
  size_t k = 1;
  while (1) {
    p = k * q + 1;
    if (prime_test(p, 10))
      break;
    k = k + 1;
  }

  // Compute elements of G_q = {h^k mod p | h in Z_p^*}
  // |G_q| = q ,G_q is the unique q-order subgroup of Z_p.
  ZZ m(1);
  // std::unordered_set<ZZ> S;
  // FIXME: unordered_set is enough
  std::set<ZZ> S;
  for (ZZ h(1); h < p; ++h) {
    S.insert(PowerMod(h, k, p));
    // G.emplace_back(tmp % p);
  }
  // rm dup
  vector<ZZ> G(S.begin(), S.end());
  std::sort(G.begin(), G.end());

  // random non-1 elements
  std::vector<ZZ> R;
  ZZ g;
  while (1) {
    std::sample(G.begin(), G.end(), std::back_inserter(R), 1,
                std::mt19937{std::random_device{}()});
    assert(R.size() != 0);
    g = R[0];
    if (g != 1)
      break;
  }

  cout << "paremeters:" << endl;
  cout << "\tq: " << q << endl;
  cout << "\tp: " << p << endl;
  // cout << "\tk: " << k << endl;
  cout << "\tg: " << g << endl;
  cout << "\tv: " << v << endl;
  // cout << "\tnum of shares: " << n << endl;
  // cout << "\tthreshold: " << t << endl << endl;
  cout << "order of G is " << G.size() << endl;

  // share to n parts
  vector<Share> shares;
  vector<ZZ> comms;
  vss_share(shares, comms, v, n, t);
  cout << "generate " << n << " shares:" << endl;
  print_shares(shares);

  // validation of shares
  vss_verify(shares, comms, n, t);

  // random sample `t` shares
  vector<Share> t_shares;
  sample(shares.begin(), shares.end(), std::back_inserter(t_shares), t,
         std::mt19937{std::random_device{}()});

  cout << "sample " << t << " shares" << endl;
  print_shares(t_shares);

  // resume to origin value
  ZZ out;
  vss_recon(out, t_shares);
  cout << "resume: " << out << endl;
  cout << "origin: " << v << endl;
}
